### ISSUE TITLE
"try calling implementing the comparison without allocating"

### ISSUE THREAD

**Issue by oli-obk** [rust-lang-nursery/rust-clippy#3289]
No idea what this is trying to tell me. Probably wants to say

"try implementing the comparison without allocating", but we should additionally look at the existing errors in our test suite and see whether we could give hints as to *how* to do this (maybe even with a suggestion).

**Comment by JoshMcguigan** [comment]
The details of this lint are fresh in my mind from #3287. 

Changing `try calling implementing the comparison without allocating` to `try implementing the comparison without allocating` is a definite improvement but doesn't solve the full problem as you noted. 

As best I can tell, the following block of code (which is where that quote comes from) was intended to avoid an incorrect suggestion for one of the test cases.

https://github.com/rust-lang-nursery/rust-clippy/blob/master/clippy_lints/src/misc.rs#L550-L562

https://github.com/rust-lang-nursery/rust-clippy/blob/master/tests/ui/cmp_owned.rs#L44

Without that `if` in the impl, clippy would suggest removing the `.to_owned` in the above test case, which would be incorrect. The trouble is, that same scenario can be triggered in such a way that it doesn't get caught by that `if`, as shown in the playground below.

https://play.rust-lang.org/?gist=a0d425f8fe25cdb34c44a38aeccca730&version=stable&mode=debug&edition=2015

I don't really understand this comment before that `if` statement. I'm guessing that this was implemented to perhaps stop clippy from panicing on some crate? If not, I think the best was to resolve this is to remove that if statement completely and then correct the resulting suggestion (after removing the `if`, clippy would suggest removing the `to_owned` but it should also suggest removing the `*` from the right side of the `==`. 

https://github.com/rust-lang-nursery/rust-clippy/blob/master/clippy_lints/src/misc.rs#L544-L547

Does that analysis seem correct?

**Comment by oli-obk** [comment]
Ha, good analysis.

Yea, we don't want to suggest removing `to_owned` inside `PartialEq` impls, as that can likely cause the suggested code to recurse on itself (implementing PartialEq by calling PartialEq itself).

Your playground test does not trigger the mentioned code. It looks totally fine.

So this issue just needs an improvement of the error message.

**Comment by JoshMcguigan** [comment]
Running the playground below creates this message from clippy for me, which is incorrect. Do you want to handle this as a separate issue? Or am I still missing something? 

https://play.rust-lang.org/?gist=a0d425f8fe25cdb34c44a38aeccca730&version=stable&mode=debug&edition=2015

```
warning: this creates an owned instance just for comparison
 --> src/main.rs:5:5
  |
5 |     y.to_owned() == *x;
  |     ^^^^^^^^^^^^ help: try: `y`
  |
  = note: #[warn(clippy::cmp_owned)] on by default
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#cmp_owned
```

**Comment by oli-obk** [comment]
Oh that is independent of this issue :laughing: but you can handle it in one PR if you want
